Backport some compiler performance improvements [ci: last-only]

retronym opened the pull request:
Backports:

  - #7109
  - #7107
  - #7105
  - #7103
  - #7102
  - #7098
  - #7097
  - #7096
  - #6990
  - #6967
  - #6607
  - #6664
  - #6631
  - #6622
  - #6618
  - #6607
  - #6605
  - #6604
  - #6599
  - #6597
  - #6524
  - #6297
  - #6292
  - #6291
  - #6234
  - #6214
  - #6209
  - #7128
  - #7149 
  - #7143

retronym committed:
Remove some indirections in Global and Typer

This reduces the number of Java stack frames between the bottom
of the stack and the start of typer, and, more importantly,
at each level of recursion in the AST or in symbol completion.

The intent is to make the stacks easier to visually scan in
profilers and other tools that display them. I'm not expecting
that performance will improve, the JVM probably does a decent
jobs and inlining these chunks of the stack.

(cherry picked from commit 3d69134a50cc1e010e135377c54a7375db13ddb9)

retronym committed:
Optimize IndexedSeqOptimized.toList

Notably, this will be used in `List(a, b, c)`.

retronym committed:
Avoid nonEmpty in hot paths

retronym committed:
Use AnyRefMap in hot parts of the compiler.

retronym committed:
Optimize nested class collection

retronym committed:
Optimize generic sig parser

retronym committed:
Simplify specialization transformer

retronym committed:
Tree convervation in specialization

retronym committed:
Optimize bookkeeping in specialization transform

retronym committed:
Reduce overhead of enabling -Ystatistics

The implementation trick of using an AlmostFinalValue to have zero
cost for the "isEnabled" check in the common case has a small flaw:
the switchpoint is tripped _every_ time stats is enabled, rather
than just on the first time. This discards a swathe of JIT compiled
code each time a Global is started with `-Ystatistics`.

This commit avoids tripping the switchpoint redundantly.

Performance:

```
⚡ for extra in "-Ystatistics:_" ""; do for v in 2.12.5-bin-91649d1-SNAPSHOT 2.12.4; do echo $v $extra; sbt 'set scalaVersion in compilation := "'$v'"' 'hot -psource=scalap -f1 -wi 5 -i 3 -pextraArgs='$extra | egrep 'HotScalacBenchmark.compile\s'; done; done
2.12.5-bin-91649d1-SNAPSHOT -Ystatistics:_
[info] HotScalacBenchmark.compile                          a8c43dc  -Ystatistics:_       false    scalap  sample   33   973.523 ± 23.389  ms/op
2.12.4 -Ystatistics:_
[info] HotScalacBenchmark.compile                          a8c43dc  -Ystatistics:_       false    scalap  sample   12  2921.333 ± 177.831  ms/op
2.12.5-bin-91649d1-SNAPSHOT
[info] HotScalacBenchmark.compile                          a8c43dc                    false    scalap  sample   38  811.846 ± 13.436  ms/op
2.12.4
[info] HotScalacBenchmark.compile                          a8c43dc                    false    scalap  sample   38  820.814 ± 17.809  ms/op
```

There is still more overhead than I would like, and it might still make
sense to move a few stats back into the "hot" category.

retronym committed:
Avoid allocation of ClassTags in hot code

Cherry pick of 94c2d4a82a1d044b4eb59b20d35ada72e9cc7ca7

retronym committed:
Refactor typechecking of array instantiation for performance

Defining the extractor object `ArrayInstantation` in `typed1` meant
that we needed to thread a `LazyRef` through to `typedApply`
(even if we were dealing with a different AST node!).

This commit moves the extractor to `TreeInfo` after extracting the
core part that relies on `Typer.this`.

(cherry picked from commit d1c90ec3dec2bf72546edcda9e4f696882a7ec61)

retronym committed:
Avoid creation of temporary Lists of imports during Context.lookup

This is a major source of allocation pressure during typechecking.

After the refactoring, we only allocate a single cursor (which is
hopefully amenable to escape analyis), which advances the pair
of pointers out the context chain.

(cherry picked from commit 7526e4550acb672dd1cdbba603bba2ee03a249e7)

retronym committed:
Avoid wasteful loading of unreferenced, high arity function and tuple classes

retronym committed:
Rule out more implicit based on bounds of type parameters

Before:

```
implicitly[Foo[String]] BYVALmode-EXPRmode (site: value <local Test> in Test)
 |-- implicitly BYVALmode-EXPRmode-FUNmode-POLYmode-TAPPmode (site: value <local Test> in Test)
 |    \-> [T](implicit e: T)T
 |-- Foo[String] TYPEmode (site: value <local Test> in Test)
 |    |-- String TYPEmode (site: value <local Test> in Test)
 |    |    [adapt] String is now a TypeTree(String)
 |    |    \-> String
 |    \-> Foo[String]
 [search #1] start `[T](implicit e: T)T`, searching for adaptation to pt=Foo[String] (silent: value <local Test> in Test) implicits disabled
 [search #1] considering Foo.javaEnumFoo
 solving for (T: ?T)
 [adapt] [T]=> Foo[T] adapted to [T]=> Foo[T] based on pt Foo[String]
 [search #1] success inferred value of type Foo[String] is SearchResult(Foo.javaEnumFoo[String], )
 |-- [T](implicit e: T)T BYVALmode-EXPRmode (site: value <local Test> in Test)
 |    \-> Foo[String]
 [adapt] [T](implicit e: T)T adapted to [T](implicit e: T)T
```

After:

```
implicitly[Foo[String]] BYVALmode-EXPRmode (site: value <local Test> in Test)
 |-- implicitly BYVALmode-EXPRmode-FUNmode-POLYmode-TAPPmode (site: value <local Test> in Test)
 |    \-> [T](implicit e: T)T
 |-- Foo[String] TYPEmode (site: value <local Test> in Test)
 |    |-- String TYPEmode (site: value <local Test> in Test)
 |    |    [adapt] String is now a TypeTree(String)
 |    |    \-> String
 |    \-> Foo[String]
 [search #1] start `[T](implicit e: T)T`, searching for adaptation to pt=Foo[String] (silent: value <local Test> in Test) implicits disabled
 [search #1] considering Foo.stringFoo
 [search #1] success inferred value of type Foo[String] is SearchResult(Foo.stringFoo, )
 |-- [T](implicit e: T)T BYVALmode-EXPRmode (site: value <local Test> in Test)
 |    \-> Foo[String]
 [adapt] [T](implicit e: T)T adapted to [T](implicit e: T)T
 \-> Foo[String]

```

(cherry picked from commit 68799b9c73de1a5d99437fce54397980a4e8e0fe)

retronym committed:
Reduce allocation in patmat updateSubstitution

(cherry picked from commit 6be8a5474bcc77d50b3688ba75ab4bb0bf90433a)

retronym committed:
Reduce allocation patmat Logic.simplify

(cherry picked from commit 3124995fd948e4a4877a2345a530b60b92b94785)

retronym committed:
Optimize patmat substitution

(cherry picked from commit d1e489ef16f66670a91139d31a17de9175e5943f)

retronym committed:
Optimize isStable

(cherry picked from commit c8fd3373c026db1b84460e2f63c9763a0c647841)

retronym committed:
Optimize isStable by using normalize, which is cached

(cherry picked from commit 9e533873f87535fc28ac7d315f04bf37697cc44e)

retronym committed:
Avoid eager error buffer creation

(cherry picked from commit 84b52194304097358479b5e5c13cc3293c36ab4f)

retronym committed:
Optimize patmat substitution

(cherry picked from commit e57bafd804120b05ac7aff2468ebd27f957bed56)

retronym committed:
Avoid an array copy in the parser

(cherry picked from commit aaf56b4bc26bfd5f4d527689e5b66f31af6e2b59)

retronym committed:
Reduce allocation further in TypedSubstitution.<<

(cherry picked from commit 9464399d2ce577b2e2ac429917d0b1f766be5a86)

milessabin committed:
Faster implicit search: lazier error messages

Don't eagerly compute unseen, yet expensive error messages
(because we're in a nested search, or -Xlog-implicits is not enabled).

(cherry picked from commit a595114cd8be644cfc5587d7ffa98710befbd8ba)

cryptom-dev committed:
Enhance performance of unapply method of ClassTag

(cherry picked from commit 25a87833345f274ad10de2b20fa0781ec88b0913)

diesalbla committed:
Solve Issue 492, "MethodTypes.isTrivial is allocation heavy"

This commit addresses Issue 492 at ScalaDev.
It aims at reducing the allocation of Cons objects in the execution of
the `isTrivial` method. In particular:

- We use an array, evaluated on demand, to contain the `_.tpe` of each
  parameter in `params`.
- We also add an array for the ContainsCollector objects, which is
  also one for each parameter. Before, the method would create one
  for each type, which meant N^2 creations in the worst case.
- We turn the auxiliary methods for `isTrivial` from recursive into
  imperative loops, using when possible indexes over those arrays.

(cherry picked from commit 9875c12a4c05b337b958e1c35590fa111739f5ed)

retronym committed:
Specialize hasImpureAtom for common cases of small collections

(cherry picked from commit fbb9d9088c47e8f6b3f9bbb9a3363afa38d8e944)

retronym committed:
Avoid needless LUB of the cases after patmat translation

I've restricted the change to the non-CPS world, where we can't assume
that Any is a top type.

(cherry picked from commit c1e81721e3c7038c17207499b79c86a63d1920ba)

diesalbla committed:
Move fields into local variables.

The method `areTrivialParams` is assumed to be used at most once
per MethodType object, so there is no need to keep the arrays as
fields. Instead, we can turn them into local method variables.

- Since the main loop goes parameter by parameter, we need not
  to keep the ContainsCollector objects, only the types.

- We unroll the main loop, so that we first check the `.isTrivial`
  method of each type, and only afterwards check with anyContains.

(cherry picked from commit 2fcf0ecbdeff6a99a9a873742ae7df237a115beb)

retronym committed:
Avoid allocation of Typer within typing transformers

Rather than allocating a new Typer each time, we can just
mutate/restore `Typer.context`.

(cherry picked from commit f7f85f2d71928c0fc045d422c8c20e49eb7b8c58)

retronym committed:
Reduce allocations in Context.lookup

This avoids allocations in the case where only a single result
is found. When more than once is found, we do still allocate
a ListBuffer as we build up the list of overloaded alternatives.

(cherry picked from commit 4eb551a1c0ed1271d90fe0c384af3b6434b6d32c)

retronym committed:
Expunge the transformed field from Typer

(cherry picked from commit 708f5990381f149564d96b47abe3a432345f08f2)

retronym committed:
Jettison runDefinitions field from Typer

(cherry picked from commit 4861fd20fe8cfd7a7958b8a551b266fee79e5fb7)

retronym committed:
Run another runDefinitions field out of town

(cherry picked from commit a431c6cd3697977bd99d45bd60eb255380656a8a)

retronym committed:
Begone, toOrigin, approximateAbstracts, and AdjustedTypeArgs!

(cherry picked from commit 37b3892fa06cf30e0335397a2d020ea757309c37)

retronym committed:
Send Typer.checkUnused to the glue factory

(cherry picked from commit 597cd9e301193e7bd95970365bd3743cc109ab3c)

retronym committed:
One context field ought to be enough for Typer, don't you think?

(cherry picked from commit 447e9f71c592d82318204119857ad6efde40cdf5)

retronym committed:
Improve efficiency of dead code checks

  - Move the checkDead module out of Typer
  - Use a new bit in ContextMode to track when to suppress
    the warning, rather than maintaining the stack of the
    symbols of enclosing applications
  - Only do any of this when under -Ywarn-dead code

References scala/scala-dev#501

(cherry picked from commit fc72bbbc7943099ab76a7e59f14ddb55f41abca3)

retronym committed:
Remove the checkNoEscaping field from Typer

(cherry picked from commit 4d28256098169a36765162fbdb06f3b8ab14fd14)

retronym committed:
Remove the dyna field from Typer

(cherry picked from commit 1cacc13d0a7d5a65d311a967277a06f886297030)

som-snytt committed:
CommandLineParser improvements

Tokenizing used to snip the line and knit the args,
this commit denits the deknitting.

Adds a benchmark showing linear behavior and a unit test
to show correctness.

Departing from the previous notion of correctness,
internal quotes are respected. `"abc"xyz` is `abcxyz`.

(cherry picked from commit 5df37eab5d485b7e95771f657e5e11bcca166e90)

retronym committed:
Rework named/defaults bookkeeping as a tree attachment

This saves a field in Context.

(cherry picked from commit 23cc067e4046baa8f2b2e1e12c27ee7becae217b)

retronym committed:
Use faster way to lookup List.apply symbol

I accidentally changed this to use Definitions.List_apply
in #6618.

(cherry picked from commit 1379df4c0888236dbc6d964d8a889b8bd9857511)

retronym committed:
Minor style changes and remove dead code

(cherry picked from commit 1581c8405e9a78397b5cf4c78567d85e9da60529)

retronym committed:
Avoid double open when overwriting a classfile on non-Windows

(cherry picked from commit 1ceafd2275f6fab4016173a4fb93dbce941805e4)

retronym committed:
Cherry-picked into a different branch

retronym committed:
Merge branch 'perf-backport/6214-optimizations-in-the-specialization-phase' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6234-reduce-overhead-of-ystatistics' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6291-avoid-allocation-of-classtags-in-hot-code' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6292-array-instantiation' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6297-imports-context-lookup' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6524-implicit-prune-bounds' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6597-front-end' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6599-allocation-patmat' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6604-classtag-unapply-newarray' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6605-methodtype-istrivial' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6607-needles-lub-match' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6612-faster-implicit-search-lazy-msg' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6615-efficiency-scope-lookup' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6616-allocation-typing-transformer' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6617-dead-code' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6618-reduce-bulk-typer-inferencer' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6622-command-line-parser' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6631-footprint-context' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6664-list-apply-rundefinitions' into perf-backport/omnibus

retronym committed:
[nomerge] Remove test that doesn't apply on 2.12.x

retronym committed:
Avoid needless storage/lookup for constants in Scope

retronym committed:
Speedup some symbol lookups in Definitions

retronym committed:
Avoid some virtual call overhead to get to Symbol.equals

retronym committed:
Use eq in lookupEntry

(cherry picked from commit ef3d9f384b848e3fab03f54a1233fa3f7b1685be)

retronym committed:
Avoid a virtual call for Phase.erasedTypes

I can't quite untangle the history to know why things are setup
this way (maybe it goes back to the MSIL backend?).

But with this small refactor we can avoid the virtual call
overhead.

retronym committed:
Move Shadower to outer level

I want to get rid of this altogeher, but in the meantime we can
be more efficnet and overhead by unnesting NoShadower.

retronym committed:
Make Name.start final

(cherry picked from commit 31f6afbb32a5e3491e09fafe4b6e35237a7feb4d)

retronym committed:
Avoid List.equals in hot path

(cherry picked from commit 4da0de224c52692c5e2374732b1ccd6b53c27009)

retronym committed:
Avoid double call to `dealias`

(cherry picked from commit c602727a6c2b7adac97149f28e6fb9cc5404391a)

retronym committed:
Avoid repeated calls to Symbol.info

(cherry picked from commit ab66e2618c4bc6c3fb6c335ff4f63edb6514b3df)

retronym committed:
Avoid expensive call to `imports` in implicit search

(cherry picked from commit 2fae7d814bc9aa77a699c30c5940d16f82e476c8)

retronym committed:
Avoid SeqFactory.unapplySeq in hot pattern matches.

(cherry picked from commit 85d8ed8408d02c662819a9d58f62beeb1c0768d0)

retronym committed:
Remove hot assertion

We want this method to be a tiny as possible so it is
JVM inlined into call sites and free of any overhead.

(cherry picked from commit 0586e022ccf736ad70c499a0939854f2176fbda4)

retronym committed:
Add a test to show a problem in switch pattern translation

The performance optimization in #6607 assumed that a translated
match would always be a `Match` node itself, but it can also be
a `{ synthetic val x1 = ...; x1 match { .. } }` block.

retronym committed:
Fix regression in switch pattern translation

The performance optimization in #6607 assumed that a translated
match would always be a `Match` node itself, but it can also be
a `{ synthetic val x1 = ...; x1 match { .. } }` block.

retronym committed:
Merge branch 'perf-backport/6607-needles-lub-match' into perf-backport/omnibus

retronym committed:
Optimize MethodType.resultType

InstantiateDependentMethodType typically doesn't encounter singleton
types, even after the initial fast path for trivial MethodTypes in
resultType

This commit defers some collection copying until the first time it
is needed, and also switches to using an Array rather than a Vector.

retronym committed:
Avoid temporary strings during classpath lookup

Index the JAR metadata by dotted package name to avoid
converting a dotted name to path on each lookup.

(cherry picked from commit 6b3f3203b5696bc3b21df6f26bd1c394e84641d6)

retronym committed:
Optimize Constant.{equals,hashCode}

 - Avoid boxing of the raw bits of double/floats before using them
   in equals/hashcode
 - Avoid cooperative equality by directly calling .equals / .hashCode
   for other values.

retronym committed:
Optimize caseFieldAccessors to avoid temporary object creation

retronym committed:
Merge branch 'perf-backport/6967-needless-storage-scope' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/6990-microopts-definitions' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/7096-optimize-method-type-result' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/7097-case-field-accessors' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/7098-constant-equals-hash' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/7102-erased-types' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/7103-shadower-outer' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/seq-unapply' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/7105-remove-indirection-typer' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/7107-imports-implicits' into perf-backport/omnibus

retronym committed:
Merge branch 'perf-backport/7109-pocketful-microopts' into perf-backport/omnibus

retronym committed:
Optimize check for enabled statistics by specializing machinery to boolean flags

Rather than returning a `ConstantCallSite` of an object that wraps a true/false
value, just directly put the boolean in the callsite.

AFAICT this eliminates a cast and a pointer dereference, even in C2 optimized
code.

(cherry picked from commit 12ae86b6a92fdcf09db83de15ac18987c926c43f)

retronym committed:
Optimize TypeRef.equals with fast paths for eq elements

(cherry picked from commit 519a320622c7c255137e5925e17a7715bafb2176)

retronym committed:
Address binary incompatibilities

  - Restore default implementation of ClassTag.apply in 2.12.x
  - Whitelist the change to add a concrete implementation of toList
    in IndexedSeqOptimized.

I tested the latter change for potential breakage with:

```
object Test {
  def main(args: Array[String]) {
    def s = new collection.immutable.WrappedString("")
    s.toList
    (s: collection.IndexedSeqOptimized[Char, Any]).toList
    (s: collection.GenTraversableOnce[Char]).toList
  }
}

```

```
$ qscalac sandbox/test.scala && scala-launch 2.12.0 Test
$ scalac-launch 2.12.0 sandbox/test.scala && qscala Test
```

retronym committed:
Use dotted keys in dirs map in other use sites of getDir

(cherry picked from commit 11230a8f247277e386ff593287a8ca073e76cd48)

retronym commented:
Benchmarks [scheduled](https://scala-ci.typesafe.com/view/scala-bench/job/compiler-benchmark/1822/), results to appear [here](https://scala-ci.typesafe.com/grafana/dashboard/db/pr-validation?orgId=1&from=1501699042299&to=1535585531323&var-scalaSha=e164092%7C4c9780fb7a%7C9a2219405b%7Cf57113ebe6%7Cf7fcf3f908%7C4779c4298c%7C3f651c3916%7C69f32c025d%7C90c94cc2ea%7C9a8328e468%7Cf552875c17&var-source=All&var-bench=HotScalacBenchmark.compile&var-host=scalabench%40scalabench%40)

retronym committed:
Merge remote-tracking branch 'origin/2.12.x' into perf-backport/omnibus-first-pr

mkeskells commented:
📣

milessabin commented:
@retronym a word of warning on https://github.com/scala/scala/pull/7147/commits/d0265db8d0ae2c7fa26cfa704570a121201f54cc ... this broke a few things in the 2.13.x community build as discussed [here](https://github.com/scala/scala/pull/6612#issuecomment-402504660). Clearly a bug is being relied on in those cases, nevertheless, including this change will break code which currently compiles on 2.12.6.

If you want to get 95% of the positive effect of that change, then https://github.com/scala/scala/pull/7067 will do it ... if you recall, early pruning eliminated most of the unnecessary thunks.

retronym commented:
Thanks, @milessabin.

I've whittled the Specs2 thing down to:

```scala
import org.specs2.matcher.{Matcher, ValueCheck, OptionLikeCheckedMatcher, OptionLikeMatcher}


trait Disj[A, B]
class Repro {
  def returnValue[T](check: ValueCheck[T]): Any = null
  implicit def matcherIsValueCheck[T](m: Matcher[T]): ValueCheck[T] = ???

  def overloaded[T](t: ValueCheck[T]): LeftDisjunctionCheckedMatcher[T] = null
  def overloaded[T]: OptionLikeMatcher[({type l[a]= Disj[a, _]})#l, T, T] = null
  trait LeftDisjunctionCheckedMatcher[T] extends OptionLikeCheckedMatcher[({type l[a]=Disj[a, _]})#l, T, T]

  returnValue(overloaded(null: Matcher[AnyRef]))
}
```

Which compiles with 2.12.6, fails with this PR, and works again with this PR with `-Xlog-implicits` which disables the error message optimization.

The `-Xlog-implicits` output is:


```
/Users/jz/code/specs2/scalaz/shared/src/main/scala/org/specs2/matcher/TaskMatchers.scala:13: matcherIsValueCheck is not a valid implicit value for Repro.this.LeftDisjunctionCheckedMatcher[=?T] => org.specs2.matcher.ValueCheck[?] because:
no type parameters for method matcherIsValueCheck: (m: org.specs2.matcher.Matcher[T])org.specs2.matcher.ValueCheck[T] exist so that it can be applied to arguments (Repro.this.LeftDisjunctionCheckedMatcher[?T])
 --- because ---
no unique instantiation of type variable T could be found
  returnValue(overloaded(null: Matcher[AnyRef]))
             ^
/Users/jz/code/specs2/scalaz/shared/src/main/scala/org/specs2/matcher/TaskMatchers.scala:13: matcherIsValueCheck is not a valid implicit value for Repro.this.LeftDisjunctionCheckedMatcher[=?T] => org.specs2.matcher.ValueCheck[?] because:
no type parameters for method matcherIsValueCheck: (m: org.specs2.matcher.Matcher[T])org.specs2.matcher.ValueCheck[T] exist so that it can be applied to arguments (Repro.this.LeftDisjunctionCheckedMatcher[?T])
 --- because ---
no unique instantiation of type variable T could be found
  returnValue(overloaded(null: Matcher[AnyRef]))
             ^
```

I think the implicit search is during static overload resolution.

I'll keep pulling the thread to see whats at fault here.

retronym commented:
@milessabin Your assessment was spot on, the use of the `=:=` type operations in `explainVariance` was registering a constraint on a type variable in the `pt` of an implicit search.

We can make the test above fail by applying this patch to 2.12.6.

```diff
$ git diff head
diff --git a/src/compiler/scala/tools/nsc/typechecker/Implicits.scala b/src/compiler/scala/tools/nsc/typechecker/Implicits.scala
index de801a3a91..9536ec658c 100644
--- a/src/compiler/scala/tools/nsc/typechecker/Implicits.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Implicits.scala
@@ -969,7 +969,10 @@ trait Implicits {
               }
             )

+          val mark = undoLog.log
           val typedFirstPending = typedImplicit(firstPending, ptChecked = true, isLocalToCallsite)
+          if (typedFirstPending.isFailure && settings.isScala213)
+            undoLog.undoTo(mark)

           // Pass the errors to `DivergentImplicitRecovery` so that it can note
           // the first `DivergentImplicitTypeError` that is being propagated
```

retronym committed:
[mergeforward] Avoid leaking constraints between typechecking implicit candidates

  - Predicate the backport of the optimization to avoid expensive
    type error creation to -Xsource:2.13, because side effects in
    explainVariance can be witnessed in type variables in the pt
    of an in-progress implicit search.
  - Fix this side-effect leakage with judicious use of undoLog
    around implicit candidate type checking, again under
    -Xsource:2.13 for bug compatibility.

I haven't managed to get a standalone test case, I've been using Specs2
with:

```
import org.specs2.matcher.{Matcher, ValueCheck, OptionLikeCheckedMatcher, OptionLikeMatcher}

trait Disj[A, B]
class Repro {
  def returnValue[T](check: ValueCheck[T]): Any = null
  implicit def matcherIsValueCheck[T](m: Matcher[T]): ValueCheck[T] = ???

  def overloaded[T](t: ValueCheck[T]): LeftDisjunctionCheckedMatcher[T] = null
  def overloaded[T]: OptionLikeMatcher[({type l[a]= Disj[a, _]})#l, T, T] = null
  trait LeftDisjunctionCheckedMatcher[T] extends OptionLikeCheckedMatcher[({type l[a]=Disj[a, _]})#l, T, T]

  returnValue(overloaded(null: Matcher[AnyRef]))
}

```

Which fails to compile with the bug fix:

```
/Users/jz/code/specs2/scalaz/shared/src/main/scala/org/specs2/matcher/TaskMatchers.scala:13: error: type mismatch
  returnValue(overloaded(null: Matcher[AnyRef]))
                             ^
```

But used to sneak through compilation before:

```
    Repro#8130.this.returnValue#15559[Disj#8127[AnyRef#1934, _]](Repro#8130.this.matcherIsValueCheck#15562[Disj#8127[AnyRef#1934, _]](Repro#8130.this.overloaded#15565[AnyRef#1934](Repro#8130.this.matcherIsValueCheck#15562[AnyRef#1934]((null: org#15.specs2#6588.matcher#6594.Matcher#7354[AnyRef#1934])))))
  }
}

warning: there was one feature warning; re-run with -feature for details
one warning found
```

retronym commented:
FTR, here's my failed attempt to fully isolate a test case:

```
trait Diffable[T]
object Diffable {
  implicit def universal[T]: Diffable[T] = null
}
class Matcher[-T]
trait ValueCheck[T]
trait BeEqualTypedValueCheck[T] extends ValueCheck[T]
object ValueCheck {
  implicit def typedValueCheck[T](expected: T)(implicit evidence$1: Diffable[T]): BeEqualTypedValueCheck[T] = null
}
trait OptionLikeCheckedMatcher[F[_], T, U] extends Matcher[F[T]]
trait OptionLikeMatcher[F[_], T, U] extends Matcher[F[T]]

trait Disj[A, B]
class Repro {
  def returnValue[T](check: ValueCheck[T]): Any = null
  implicit def matcherIsValueCheck[T](m: Matcher[T]): ValueCheck[T] = ???

  def overloaded[T](t: ValueCheck[T]): LeftDisjunctionCheckedMatcher[T] = null
  def overloaded[T]: OptionLikeMatcher[({type l[a]= Disj[a, _]})#l, T, T] = null
  trait LeftDisjunctionCheckedMatcher[T] extends OptionLikeCheckedMatcher[({type l[a]=Disj[a, _]})#l, T, T]

  returnValue(overloaded(null: Matcher[AnyRef]))
}

```

retronym committed:
Merge branch 'perf-backport/7128-temp-string-classpath' into perf-backport/omnibus-first-pr

retronym committed:
Merge branch 'perf-backport/7149-cold-stats' into perf-backport/omnibus-first-pr

retronym committed:
Merge branch 'perf-backport/7143-type-ref-equals' into perf-backport/omnibus-first-pr

retronym committed:
Address binary incompatibility

Refactor to make one new method private, and make the other
new method package private with a whitelist exclusion on the
grounds that scala-reflect and scala-compiler JARs are expected
to be identically versioned.

retronym commented:
Added a few more, which will add to the [extended benchmark plot](https://scala-ci.typesafe.com/grafana/dashboard/db/pr-validation?orgId=1&from=1501699042299&to=1535585531323&var-scalaSha=5d34c2dfab%7C0e901ba834%7C87bcc75005%7C6be348ac8d%7C6ea993a090%7C03feb56902%7Ce164092a73%7Ce46c4ba144%7Ca857be2e12%7Cb248c1d654%7C2cd6133323%7C346b7ea6a8%7Cad9820d4ff%7C492201babf%7C1a09b1574d%7C799ab6c071%7Cb18f1ac81c%7C00673bbf8f%7C4a55ac81b1%7Cf3c70d9376%7Ce59db3921e%7C81d6b754e5%7Cc5eb150b18%7C1e8480cacf%7C0e579bc13e%7Cf6c5093547%7C798a04e84c%7C4ad2edc4a4%7Cec4846288e%7Cf0c2db828c%7C3ac6c1e422%7Cbe571b66a2%7Cc9f04d0237%7C940e3c2567%7C15f13f1c9e%7Ccc939a52d0%7C1561e1399a%7Ca4fee32dd2%7C6c0b70a81e%7C39a567b8de%7C8398ca5c23%7C833cf7ef13%7C2062b3e63c%7C17fb78c1bf%7Caac607b81f%7Cbd76826ab8%7Caae6a977aa%7Cc1b68e4f4b%7C11ef06f030%7C9e8a421586%7Cc15a386e33%7C32f1944084%7Cd508e7df95%7C9136c5011b%7C90c5264716%7Ce2a0bd9925%7Cc617272912%7Cc3f53687d1%7Cdb73686aaf%7C8a7e43922f&var-source=All&var-bench=HotScalacBenchmark.compile&var-host=scalabench%40scalabench%40)

retronym committed:
[mergeforward] Make nested implicit type error suppression unconditional

Similarly, make the type constrain undo bugfix unconditional.

This is known to break specs2, I'd like to see if that's all we
find.

https://github.com/scala/scala/pull/7147#issuecomment-418233611

retronym committed:
Merge remote-tracking branch 'origin/2.12.x' into perf-backport/omnibus-first-pr

retronym commented:
/synch

retronym commented:
@adriaanm The main thing to review here is the `undoLog` bug fix to `Implicits`, which was motivated by the investigation of the way that the side effects leak was causing specs2 compilation to change course when error message creation was surpressed.

I started out predicating this on `-Xsource:2.13`, but added a commit to turn the fix on unconditionally after we chatted about this and thought it might be such a corner case that we'd be willing to sacrifice bug-compatibility.

retronym commented:
I wasn't able to identify subset of commits that would give most of the performance benefit, performance ratchets up fairly steadily.

```
c15a386e33fd7ba1b7c1b8635cf9d09d3c6c723f,    838.13,Merge pull request #6697 from lrytz/code-mono
9e8a4215863db97eca7f3f18ae16a88931793934,    828.10,Merge pull request #6655 from hrhino/t10886
11ef06f03048cef40ec3d3210c77a61058fd0a01,    831.65,Merge pull request #6672 from hrhino/t10880
c1b68e4f4bf17be5aaaa2f008d271f26983096ca,    823.90,Merge pull request #6470 from DavidGregory084/bug#10808
aae6a977aadc67b53e75972190327570185fbe03,    835.85,Merge pull request #6439 from som-snytt/issue/10784
bd76826ab8786ce6fd344b4d7eace972bf85cc3c,    828.80,Merge pull request #6712 from exoego/scaladoc-placeholder-color
aac607b81ff5ad9f4402ef2121782cf97be9d4a3,    827.86,Merge pull request #6596 from exoego/older-version-notice
17fb78c1bfa34777cfdaeca0ec65c485a84f96c1,      0.00,Optimize IndexedSeqOptimized.toList
2062b3e63c70d6fad0c87a3c8d317525577441b2,      0.00,Use AnyRefMap in hot parts of the compiler.
833cf7ef13bf65877c3730c41f3fea63468ff863,    827.50,Optimize nested class collection
8398ca5c23a79ac7175e2e0382a1e3e74f695546,    831.60,Optimize generic sig parser
39a567b8defa7eb3f7593774b0dbdd68023825fd,    830.61,Avoid nonEmpty in hot paths
6c0b70a81e79ba5d535aaaa9b1ecfdd714a6ede0,    808.66,Merge branch 'perf-backport/6214-optimizations-in-the-specialization-phase' into perf-backport/omnibus
a4fee32dd2dbf0cbc9b26f170a2c1d160c7f04bc,    802.64,Merge branch 'perf-backport/6234-reduce-overhead-of-ystatistics' into perf-backport/omnibus
1561e1399a384c7067289b73b8d5ebeced588ea5,    808.37,Merge branch 'perf-backport/6291-avoid-allocation-of-classtags-in-hot-code' into perf-backport/omnibus
cc939a52d0ef308e210540867fdbab7a74c95441,    811.80,Merge branch 'perf-backport/6292-array-instantiation' into perf-backport/omnibus
15f13f1c9e887773d91d1884f4ab5891362d67e9,    808.46,Merge branch 'perf-backport/6297-imports-context-lookup' into perf-backport/omnibus
940e3c256721386585dfb96b1f29734b7f2851ab,    813.61,Merge branch 'perf-backport/6524-implicit-prune-bounds' into perf-backport/omnibus
c9f04d023774cc2765a8e16f4593a783de261f80,    804.88,Merge branch 'perf-backport/6597-front-end' into perf-backport/omnibus
be571b66a2b637dea3be239cff2189787abe3e8c,    798.59,Merge branch 'perf-backport/6599-allocation-patmat' into perf-backport/omnibus
3ac6c1e42212566985838104bdc6ff0d459fee57,    790.17,Merge branch 'perf-backport/6604-classtag-unapply-newarray' into perf-backport/omnibus
f0c2db828c202c1c3ef52d64b227837fdcd346c5,    796.63,Merge branch 'perf-backport/6605-methodtype-istrivial' into perf-backport/omnibus
ec4846288e1ec2f750220509c3fc98a8e39ab59d,    791.20,Merge branch 'perf-backport/6607-needles-lub-match' into perf-backport/omnibus
4ad2edc4a42077852a4f13eecfd864ee202a940a,    798.15,Merge branch 'perf-backport/6612-faster-implicit-search-lazy-msg' into perf-backport/omnibus
798a04e84cd1d2bbb2e61e8581cb6e66e248080a,    789.96,Merge branch 'perf-backport/6615-efficiency-scope-lookup' into perf-backport/omnibus
f6c5093547f3d1e021499cf5fb90293045519d49,    792.84,Merge branch 'perf-backport/6616-allocation-typing-transformer' into perf-backport/omnibus
0e579bc13e35ed74bb353199c90435b807792fcd,    783.00,Merge branch 'perf-backport/6617-dead-code' into perf-backport/omnibus
1e8480cacf109e99c8824b67b3c94503ac95fbce,    777.84,Merge branch 'perf-backport/6618-reduce-bulk-typer-inferencer' into perf-backport/omnibus
c5eb150b186c3c1fc4de8c19c230538926cc0571,    770.31,Merge branch 'perf-backport/6622-command-line-parser' into perf-backport/omnibus
81d6b754e5685a5bdb405f6d658ec9edf182412f,    781.56,Merge branch 'perf-backport/6631-footprint-context' into perf-backport/omnibus
e59db3921e19a5c5d4c9326964279e408dfaf8d0,    778.99,Merge branch 'perf-backport/6664-list-apply-rundefinitions' into perf-backport/omnibus
f3c70d9376290b332de7b4f8225ce032e3792ebc,    775.84,[nomerge] Remove test that doesn't apply on 2.12.x
4a55ac81b1a1ca5125fac52e7434d741fa96c787,    767.77,Merge branch 'perf-backport/6607-needles-lub-match' into perf-backport/omnibus
00673bbf8f9e01d3a051890366100ffe7e79b25e,    776.42,Merge branch 'perf-backport/6967-needless-storage-scope' into perf-backport/omnibus
b18f1ac81c58fb8aa1af8c5531ae3885ba866af7,    767.58,Merge branch 'perf-backport/6990-microopts-definitions' into perf-backport/omnibus
799ab6c0710db7dfeee38e058ff03b241a91905f,    762.45,Merge branch 'perf-backport/7096-optimize-method-type-result' into perf-backport/omnibus
1a09b1574d28684d113f7e90cd6ff1f621864297,    758.73,Merge branch 'perf-backport/7097-case-field-accessors' into perf-backport/omnibus
492201babf0cd462074852b9b52a946499344baa,    761.76,Merge branch 'perf-backport/7098-constant-equals-hash' into perf-backport/omnibus
ad9820d4ff682b0ed671116e2aa63e4a85fb6168,    766.90,Merge branch 'perf-backport/7102-erased-types' into perf-backport/omnibus
346b7ea6a85c309570a37313f392f4f1c62747ab,    750.97,Merge branch 'perf-backport/7103-shadower-outer' into perf-backport/omnibus
2cd613332308667e41637a5df559f2502781c6ee,    754.60,Merge branch 'perf-backport/seq-unapply' into perf-backport/omnibus
b248c1d654746ec0a86ae94a7166b00713aaf2ed,    759.32,Merge branch 'perf-backport/7105-remove-indirection-typer' into perf-backport/omnibus
a857be2e121607ae9609c8c6f5578f709d98e498,    763.12,Merge branch 'perf-backport/7107-imports-implicits' into perf-backport/omnibus
e46c4ba14448d841534eacd71de8369bb507065d,    750.16,Merge branch 'perf-backport/7109-pocketful-microopts' into perf-backport/omnibus
e164092a73835b170a6a706f7180008fb1793eb9,    755.62,Address binary incompatibilities
03feb5690248456c2a2db924a4cc32ee27e6f149,    753.98,Merge remote-tracking branch 'origin/2.12.x' into perf-backport/omnibus-first-pr
6ea993a090a34e1c5c53ea38b2c416ae4d1c5d36,    758.05,[mergeforward] Avoid leaking constraints between typechecking implicit candidates
6be348ac8d8ea3ff2fff7ee6d47d6b09de5f0783,      0.00,Merge branch 'perf-backport/7128-temp-string-classpath' into perf-backport/omnibus-first-pr
87bcc750059e0c570dd3ef6166299880fa15f7fd,      0.00,Merge branch 'perf-backport/7149-cold-stats' into perf-backport/omnibus-first-pr
0e901ba834c8c1b8d396c4f04e07b6f7ce576b3b,      0.00,Merge branch 'perf-backport/7143-type-ref-equals' into perf-backport/omnibus-first-pr
5d34c2dfab4ac086079cc5b312187a81a031b61f,    740.20,Address binary incompatibility
0a8e00cb1872e1c032c5f57a447743d27790ba2e,      0.00,[mergeforward] Make nested implicit type error suppression unconditional
da58d8e4f13c75f9ac757ef6f78d40b9cd885c15,      0.00,Merge remote-tracking branch 'origin/2.12.x' into perf-backport/omnibus-first-pr
```

retronym commented:
This will need a careful merge forward to 2.13.x, I'll volunteer for that job.

adriaanm commented:
/rebuild da58d8e

adriaanm commented:
Manually kicked off a build of that last sha, since the list of commits is too long for scabot (haven't implemented paging). https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-validate-main/1028/console

retronym commented:
This PR breaks `scalameta` which has a [copy/paste version of vast swathes of Typer](https://github.com/scalameta/scalameta/find/master#L12) that no longer compiles.

@olafurpg We'd need another PR like https://github.com/scalameta/scalameta/pull/1557 to prepare for 2.12.7. Is that something you're able to help out with?

This coupling is really, really undesirable.

retronym committed:
Revert change that is binary incompatible with the 2.12.6 build of macro paradise

retronym committed:
Make some methods protected that are used by scalameta's Typer subclass

adriaanm commented on src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:
```diff
@@ -201,8 +201,8 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
       exitingSpecialize {
-        FunctionClass.seq.map(_.info)
-        TupleClass.seq.map(_.info)
+        FunctionClass.seq.take(MaxFunctionAritySpecialized + 1).foreach(_.info)
```
puzzling :-)

adriaanm replied:
I understand the reason for MaxFunctionAritySpecialized thanks to the test, but why +1?

som-snytt replied:
There is a Scalafix rule now that calculates the probability of an off-by-one error and injects an adjusting `+ 1` bzw. `- 1`.

adriaanm commented:
Thanks for dealing with the macro-paradise/semanticdb fallout. Let's get this merged.
